File: ios/renderer/CodeBlockRenderer.m
#import "CodeBlockRenderer.h"
#import "CodeBlockBackground.h"
#import "LastElementUtils.h"
#import "MarkdownASTNode.h"
#import "ParagraphStyleUtils.h"
#import "RenderContext.h"
#import "RendererFactory.h"
#import "StyleConfig.h"

@implementation CodeBlockRenderer {
  __weak RendererFactory *_rendererFactory;
  StyleConfig *_config;
}

- (instancetype)initWithRendererFactory:(id)rendererFactory config:(id)config
{
  if (self = [super init]) {
    _rendererFactory = rendererFactory;
    _config = (StyleConfig *)config;
  }
  return self;
}

- (void)renderNode:(MarkdownASTNode *)node into:(NSMutableAttributedString *)output context:(RenderContext *)context
{
  [context setBlockStyle:BlockTypeCodeBlock font:[_config codeBlockFont] color:[_config codeBlockColor] headingLevel:0];

  CGFloat padding = [_config codeBlockPadding];
  CGFloat lineHeight = [_config codeBlockLineHeight];
  CGFloat marginTop = [_config codeBlockMarginTop];
  CGFloat marginBottom = [_config codeBlockMarginBottom];

  NSUInteger blockStart = output.length;
  blockStart += applyBlockSpacingBefore(output, blockStart, marginTop);

  // Top Padding: Inserted as a spacer character inside the background area
  [output appendAttributedString:kNewlineAttributedString];
  NSMutableParagraphStyle *topSpacerStyle = [context spacerStyleWithHeight:padding spacing:0];
  topSpacerStyle.baseWritingDirection = NSWritingDirectionLeftToRight;
  [output addAttribute:NSParagraphStyleAttributeName value:topSpacerStyle range:NSMakeRange(blockStart, 1)];

  NSUInteger contentStart = output.length;
  @try {
    [_rendererFactory renderChildrenOfNode:node into:output context:context];
  } @finally {
    [context clearBlockStyle];
  }

  NSUInteger contentEnd = output.length;
  if (contentEnd <= contentStart)
    return;

  NSRange contentRange = NSMakeRange(contentStart, contentEnd - contentStart);

  UIFont *codeFont = [_config codeBlockFont];
  RCTUIColor *codeColor = [_config codeBlockColor];
  if (codeColor) {
    [output addAttributes:@{NSFontAttributeName : codeFont, NSForegroundColorAttributeName : codeColor}
                    range:contentRange];
  } else {
    [output addAttribute:NSFontAttributeName value:codeFont range:contentRange];
  }

  if (lineHeight > 0) {
    applyLineHeight(output, contentRange, lineHeight);
  }

  // Code is always LTR regardless of app writing direction
  NSMutableParagraphStyle *baseStyle = [getOrCreateParagraphStyle(output, contentStart) mutableCopy];
  baseStyle.baseWritingDirection = NSWritingDirectionLeftToRight;
  baseStyle.alignment = NSTextAlignmentLeft;
  baseStyle.firstLineHeadIndent = padding;
  baseStyle.headIndent = padding;
  baseStyle.tailIndent = -padding;
  [output addAttribute:NSParagraphStyleAttributeName value:baseStyle range:contentRange];

  // Bottom Padding: Inserted as a spacer character inside the background area
  NSUInteger bottomPaddingStart = output.length;
  [output appendAttributedString:kNewlineAttributedString];
  NSMutableParagraphStyle *bottomPaddingStyle = [context spacerStyleWithHeight:padding spacing:0];
  bottomPaddingStyle.baseWritingDirection = NSWritingDirectionLeftToRight;
  [output addAttribute:NSParagraphStyleAttributeName value:bottomPaddingStyle range:NSMakeRange(bottomPaddingStart, 1)];

  // Define the range for background rendering (includes padding, excludes margins)
  NSRange backgroundRange = NSMakeRange(blockStart, output.length - blockStart);
  [output addAttribute:CodeBlockAttributeName value:@YES range:backgroundRange];

  // External Margin: Applied outside the background range
  if (marginBottom > 0) {
    applyBlockSpacingAfter(output, marginBottom);
  }
}

@end

File: packages/ios-enriched-markdown/Sources/EnrichedMarkdown/Rendering/Renderers/CodeBlockRenderer.swift
import UIKit

final class CodeBlockRenderer: NodeRenderer {
    private let factory: RendererFactory
    private let config: MarkdownStyleConfig

    init(factory: RendererFactory, config: MarkdownStyleConfig) {
        self.factory = factory
        self.config = config
    }

    func render(node: MarkdownASTNode, into output: NSMutableAttributedString, context: RenderContext) {
        let blockStyle = config.codeBlock
        let font = blockStyle.font ?? UIFont.monospacedSystemFont(ofSize: 14, weight: .regular)
        let color = blockStyle.foregroundColor ?? UIColor.label
        context.setBlockStyle(font: font, color: color, blockType: .codeBlock)

        let padding = blockStyle.padding ?? 0
        let lineHeight = blockStyle.lineHeight ?? 0
        let marginTop = blockStyle.marginTop ?? 0
        let marginBottom = blockStyle.marginBottom ?? 0

        ParagraphStyleHelpers.ensureStartingOnNewLine(in: output)

        var blockStart = output.length
        blockStart += ParagraphStyleHelpers.applyBlockSpacingBefore(
            to: output,
            at: blockStart,
            marginTop: marginTop
        )

        if padding > 0 {
            let topSpacerLocation = output.length
            output.append(ParagraphStyleHelpers.newline)
            let topSpacerStyle = ParagraphStyleHelpers.spacerParagraphStyle(height: padding)
            topSpacerStyle.baseWritingDirection = .leftToRight
            output.addAttribute(
                .paragraphStyle,
                value: topSpacerStyle,
                range: NSRange(location: topSpacerLocation, length: 1)
            )
        }

        let contentStart = output.length
        factory.renderChildren(of: node, into: output, context: context)
        context.clearBlockStyle()

        guard output.length > contentStart else { return }

        let contentRange = NSRange(location: contentStart, length: output.length - contentStart)
        let codeFont = blockStyle.font ?? UIFont.monospacedSystemFont(ofSize: 14, weight: .regular)

        if let codeColor = blockStyle.foregroundColor {
            output.addAttributes(
                [.font: codeFont, .foregroundColor: codeColor],
                range: contentRange
            )
        } else {
            output.addAttribute(.font, value: codeFont, range: contentRange)
        }

        let baseStyle = ParagraphStyleHelpers.getOrCreateParagraphStyle(in: output, at: contentStart)
        baseStyle.baseWritingDirection = .leftToRight
        baseStyle.alignment = .left
        baseStyle.firstLineHeadIndent = padding
        baseStyle.headIndent = padding
        baseStyle.tailIndent = -padding
        output.addAttribute(.paragraphStyle, value: baseStyle, range: contentRange)

        if lineHeight > 0 {
            ParagraphStyleHelpers.applyBlockLineHeight(to: output, range: contentRange, lineHeight: lineHeight)
        }

        if padding > 0 {
            let bottomSpacerLocation = output.length
            output.append(ParagraphStyleHelpers.newline)
            let bottomSpacerStyle = ParagraphStyleHelpers.spacerParagraphStyle(height: padding)
            bottomSpacerStyle.baseWritingDirection = .leftToRight
            output.addAttribute(
                .paragraphStyle,
                value: bottomSpacerStyle,
                range: NSRange(location: bottomSpacerLocation, length: 1)
            )
        }

        let backgroundRange = NSRange(location: blockStart, length: output.length - blockStart)
        output.addAttribute(MarkdownAttribute.codeBlock, value: true, range: backgroundRange)

        if marginBottom > 0 {
            ParagraphStyleHelpers.applyBlockSpacingAfter(to: output, marginBottom: marginBottom)
        }
    }
}
